e2e tests for Federated Secrets

kshafiee opened the pull request:
e2e test to confirm that underlying secrets are always created correctly in underlying clusters

This is to address Issue #29337

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31655)

<!-- Reviewable:end -->

kshafiee committed:
e2e tests for Federated Secrets

kshafiee commented:
@nikhiljindal @quinton-hoole @jianhuiz @deepak-vij

kshafiee commented:
ok to test

quinton-hoole commented:
ok to test

quinton-hoole commented:
@k8s-bot federation gce e2e test this

quinton-hoole commented:
You can ignore the "Jenkins Federation GCE e2e" error - it's unrelated to this PR.  I've filed an issue to have it fixed. https://github.com/kubernetes/test-infra/issues/470

kshafiee commented:
Thanks @quinton-hoole , sure!

quinton-hoole commented on test/e2e/federated-secret.go:
```diff
@@ -35,52 +36,111 @@ const (
+				deleteAllTestSecrets(
+					clientset.Core().Secrets(nsName).List,
+					clientset.Core().Secrets(nsName).Delete,
+					false)
```
Ideally you need a call to unregisterClusters at the end of ForEach().  See 

https://github.com/kubernetes/kubernetes/blob/8c110675406ff822e7245563f9911fb173a0015e/test/e2e/federated-ingress.go#L101

and 

https://github.com/kubernetes/kubernetes/blob/4edc98260f26b2047ae9fcfa33cca1554dd1a3b2/test/e2e/federation-util.go#L163

kshafiee replied:
I am using function 'buildClustersOrFail_14' in 'federation-util-14.go' that returns clusters of type '[]*federation_api.Cluster'
In the federated-ingress.go example Jesse is using function 'registerClusters' in 'federation-util.go' that takes clusters of type 'map[string]*cluster'

The reason I used 'federation-util-14.go' was I saw federated-namespace.go as an example. Please let me know which one I should use. Thx!

nikhiljindal replied:
Yes unfortunately we have ended up having those 2 util files that do the same thing in slightly different ways.
We want to consolidate them.

If you update this code to use federation-util.go, then feel free to make the same change to the namespace test. Ref: https://github.com/kubernetes/kubernetes/pull/30443#issuecomment-239492984.

cc @mwielgus

quinton-hoole commented on test/e2e/federated-secret.go:
```diff
@@ -35,52 +36,111 @@ const (
 
 			nsName := f.FederationNamespace.Name
-			secret := createSecretOrFail(f.FederationClientset_1_4, nsName)
+
```
Why this change?

quinton-hoole commented on test/e2e/federated-secret.go:
```diff
@@ -35,52 +36,111 @@ const (
-			err := f.FederationClientset_1_4.Core().Secrets(nsName).Delete(secret.Name, &api.DeleteOptions{})
-			framework.ExpectNoError(err, "Error deleting secret %q in namespace %q", secret.Name, secret.Namespace)
-			By(fmt.Sprintf("Deletion of secret %q in namespace %q succeeded.", secret.Name, nsName))
+			deleteAllTestSecrets(
```
You already have this in AfterEach(), so you shouldn't need it here?

quinton-hoole commented on test/e2e/federated-secret.go:
```diff
@@ -35,52 +36,111 @@ const (
+			_, err := f.FederationClientset_1_4.Core().Secrets(nsName).Create(secret)
+			framework.ExpectNoError(err, "Failed to create secret %s", secret.Name)
+
+			// Check subclusters if the secret was created there
```
nit: I would prefer this in a separate function, so that it can be more easily used in future.

quinton-hoole commented:
Thanks @kshafiee .  A few small changes, and then LGTM.  We will also need an update test (which should check that the update is visible in the federation, and also eventually in the underlying clusters).  Please file an issue for that, and if possible, start work on it.

kshafiee commented:
ok to test

quinton-hoole commented:
@kshafiee Is this ready for re-review and merge?  Let me know and I'll do it.

kshafiee commented:
Thanks @quinton-hoole for your comments!

- I updated the code to use the functions in federation-util.go
- I added an update test

It is good for re-review now!

nikhiljindal commented on test/e2e/federated-secret.go:
```diff
@@ -18,69 +18,161 @@ package e2e
+		},
+	}
+
+	for MaxRetriesOnFederatedApiserver := 0; MaxRetriesOnFederatedApiserver < 3; MaxRetriesOnFederatedApiserver++ {
```
MaxRetries could be a constant at the top alongside FederatedSecretTimeout and use a variable retryCount here?

nikhiljindal commented on test/e2e/federated-secret.go:
```diff
@@ -18,69 +18,161 @@ package e2e
-	}
-	By(fmt.Sprintf("Creating federated secret %q in namespace %q", FederatedSecretName, namespace))
-
+func createSecretOrFail(clientset *federation_release_1_4.Clientset, namespace string) (*v1.Secret, error) {
```
You seem to be ignoring the returned err everywhere assuming the function will fail if there is an error.
No need to return an error in that case?

Same for the updateSecret fn

nikhiljindal commented on test/e2e/federated-secret.go:
```diff
@@ -18,69 +18,161 @@ package e2e
+		})
+
+		AfterEach(func() {
+			unregisterClusters(clusters, f)
```
Add framework.SkipUnlessFederated(f.Client)

nikhiljindal commented on test/e2e/federated-secret.go:
```diff
@@ -18,69 +18,161 @@ package e2e
+				federationName = DefaultFederationName
 			}
+			clusters = map[string]*cluster{}
+			_ = registerClusters(clusters, UserAgentName, federationName, f)
```
registerClusters is not using the federationName. remove it?

kshafiee commented:
Thanks @nikhiljindal ! Let me know if your comments are addressed.

nikhiljindal commented on test/e2e/federated-secret.go:
```diff
@@ -18,69 +18,157 @@ package e2e
+		},
+	}
+
+	for RetryCount := 0; RetryCount < MaxRetries; RetryCount++ {
```
RetryCount -> retryCount

nikhiljindal commented on test/e2e/federated-secret.go:
```diff
@@ -18,69 +18,157 @@ package e2e
 })
 
 func createSecretOrFail(clientset *federation_release_1_4.Clientset, namespace string) *v1.Secret {
-	if clientset == nil || len(namespace) == 0 {
```
Why delete this?
Seems useful to keep.

nikhiljindal commented:
Thanks @kshafiee 
2 minor comments, mostly LGTM

kshafiee commented:
Thx @nikhiljindal comments addressed!

nikhiljindal commented:
LGTM, thanks

kshafiee commented:
ok to test

nikhiljindal commented:
@k8s-bot unit test this issue: #IGNORE

nikhiljindal commented on test/e2e/federated-secret.go:
```diff
@@ -18,69 +18,166 @@ package e2e
+		if err != nil {
+			framework.Failf("failed to get secret %q: %v", FederatedSecretName, err)
+		}
+		newSecret, err = clientset.Core().Secrets(namespace).Update(secret)
```
This command is failing with:
```
failed to update secret "federated-secret": secrets "updated-federated-secret" not found
```

We cannot change the name of a resource while updating it.

cc @kshafiee
